### PR TITLE
fix llama4 multimodal logits

### PR DESCRIPTION
**Pull request by aireenmei**
# Description

* Fix some bugs that caused logit gap in Llama4 multimodal
* Logit check confirms the Llama4 does not use bidirectional mask for image tokens in text decoder, I remove it. 
* Improve scripts for generating golden logits, and forward pass logit check
* check_llama4_layers.py with this PR's change pass before nnx migration [PR](https://github.com/AI-Hypercomputer/maxtext/pull/2269), but fail after due to some other issues related to nnx, created b/452960432 to track

b/436601539

# Tests

The changes improves logit gap, but the decode output is still the same: [Log](https://cloudlogging.app.goo.gl/oRCSxhjRXsTK4rb2A)
More results are discussed in https://buganizer.corp.google.com/issues/436601539#comment13

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: tests/forward_pass_logit_checker.py [hunk 1]
```diff
@@ -311,15 +320,22 @@ def main(config, test_args):  # pylint: disable=W0621
       )
       max_logging.log(msg)
 
-      model_probabilities = jax.nn.softmax(train_logits_slice, axis=-1)
-      golden_probabilities = jax.nn.softmax(golden_logits_slice, axis=-1)
+      epsilon = 1e-8
+      model_probabilities = jnp.clip(jax.nn.softmax(train_logits_slice, axis=-1), a_min=epsilon)
```
Comment (RissyRan) [inline tests/forward_pass_logit_checker.py]: Could we make this as optionally?

File: tests/forward_pass_logit_checker.py [hunk 2]
```diff
@@ -249,9 +247,20 @@ def main(config, test_args):  # pylint: disable=W0621
       input_golden_data_path = os.path.join(MAXTEXT_TEST_ASSETS_ROOT, f"golden_data_{config.model_name}.jsonl")
     else:
       input_golden_data_path = test_args.golden_logits_path
-    max_logging.log("loading hf goldens from jsonl file")
-    with jsonlines.open(input_golden_data_path, "r") as f:
-      golden_data = list(f)
+    if test_args.golden_logits_path.endswith(".jsonl"):
```
Comment (RissyRan) [inline tests/forward_pass_logit_checker.py]: Nit refactor (optional)

```
file_path = Path(test_args.golden_logits_path)

if file_path.suffix == ".jsonl":
    max_logging.log(f"Loading hf goldens from jsonl file: {file_path}")
    import jsonlines  # pylint: disable=import-outside-toplevel

    with jsonlines.open(file_path, "r") as f:
        golden_data = list(f)
        
elif file_path.suffix in (".pickle", ".pkl"):
    max_logging.log(f"Loading hf goldens from pickle file: {file_path}")
    import pickle  # pylint: disable=import-outside-toplevel

    with open(file_path, "rb") as f:
        golden_data = pickle.load(f)
        
else:
    raise ValueError(
        f"Unsupported file type: {file_path.suffix}. "
        "golden_logits_path must end with .jsonl, .pickle, or .pkl"
    )

```

**Review by RissyRan — COMMENTED** [review]
LGTM with minor comment

File: tests/forward_pass_logit_checker.py (hunk 1 above)
Comment (aireenmei) [reply tests/forward_pass_logit_checker.py]: Sounds good. I added a flag `--clip_logits_epsilon`, it will only clip when it's not None

File: tests/forward_pass_logit_checker.py (hunk 2 above)
Comment (aireenmei) [reply tests/forward_pass_logit_checker.py]: refactored

File: src/MaxText/layers/gemma3.py [hunk 3]
```diff
@@ -413,13 +413,19 @@ def __init__(
     self.rngs = rngs
     self.seq_len = (self.config.image_size_for_vit // self.config.patch_size_for_vit) ** 2
 
-    self.LayerNorm_0 = nnx.LayerNorm(num_features=self.config.hidden_size_for_vit, rngs=self.rngs)
+    self.LayerNorm_0 = nnx.LayerNorm(
+        num_features=self.config.hidden_size_for_vit, epsilon=self.config.normalization_layer_epsilon, rngs=self.rngs
```
Comment (shuningjin) [inline src/MaxText/layers/gemma3.py:417]: nit: could you add a comment in base.yml to help explain what this config does? (e.g., used in layernorm, rmsnorm, etc)
https://github.com/AI-Hypercomputer/maxtext/blob/9b2f657683f411c260c76dda610715f80376f711/src/MaxText/configs/base.yml#L21

**Review by hengtaoguo — APPROVED** [review]
LGTM, great fix!

File: tests/forward_pass_logit_checker.py [hunk 4]
```diff
@@ -461,6 +482,7 @@ def main(config, test_args):  # pylint: disable=W0621
   parser.add_argument("--run_hf_model", type=bool, required=False, default=False)
   parser.add_argument("--output_logits_path", type=str, required=False, default="")
```
Comment (shuningjin) [inline tests/forward_pass_logit_checker.py:484]: nit: add a comment, e.g., support file format: json with suffix ".jsonl", and pickle with suffix ".pickle" or ".pkl"

**Review by shuningjin — APPROVED** [review]
Thanks for improving the scripts and fixing the gap!

File: tests/forward_pass_logit_checker.py (hunk 4 above)
Comment (aireenmei) [reply tests/forward_pass_logit_checker.py:484]: Added

File: src/MaxText/layers/gemma3.py (hunk 3 above)
Comment (aireenmei) [reply src/MaxText/layers/gemma3.py:417]: Added
